File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # Create a dummy node to handle edge cases where the head itself is a duplicate
        dummy = ListNode(0)
        dummy.next = head
        
        # Initialize prev as dummy and current as head
        prev = dummy
        current = head
        
        while current:
            # Check if the current node is a duplicate
            if current.next and current.val == current.next.val:
                # Skip all nodes with the same value
                while current.next and current.val == current.next.val:
                    current = current.next
                # After skipping, link prev to current's next (skip duplicates)
                prev.next = current.next
            else:
                # No duplicates, move prev to current
                prev = prev.next
            # Move current to the next node
            current = current.next
        
        # Return the updated list, skipping dummy node
        return dummy.next

File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.cpp
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        // Dummy node to handle edge cases (like head being a duplicate)
        ListNode* dummy = new ListNode(0, head);
        ListNode* prev = dummy; // Pointer to the node before the current sequence
        
        while (head) {
            // If there's a duplicate (i.e., head and head->next have the same value)
            if (head->next && head->val == head->next->val) {
                // Skip all nodes with the same value
                while (head->next && head->val == head->next->val) {
                    head = head->next;
                }
                // Move prev's next to the node after the last duplicate
                prev->next = head->next;
            } else {
                // If no duplicate, move prev to the next node
                prev = prev->next;
            }
            // Move head forward
            head = head->next;
        }
        
        // Return the updated list (skipping the dummy node)
        return dummy->next;
    }
};
